ElrondNetwork/elrond-go
# EN-7479: storer 2 elastic bug fixes

Fixed the following bugs:
- rating was not indexed for genesis -> solved by fetching initial rating data from the nodesSetup file (if starting from genesis only)
- return with error when a receipts' miniblock is not found
- use round duration in format of seconds instead of milliseconds in tps benchmark updater (TPS and round time were not calculated correctly because of this difference)
- index validators in epoch 0 (if starting from genesis only)
- use `GetMiniBlockHeadersHashes` for fetching all the miniblock hashes of a header, because in the previous version, not all the miniblocks were included
- process and index all headers for a shard which are notarized in a meta block. before, only one header per shard was indexed and this raised inconsistencies as there were missing headers.
- used index hashed nodes coordinator with rating as the base node coordinator wouldn't have returned the same data regarding validators in an epoch or in a round.

Testing procedure:
1. run a testnet (with this branch or at least with `development` branch) with `elastic search` on for some epochs (but limited to maximum 10k transactions/blocks/miniblocks so our comparison tool will work properly)
2. on a machine with this branch on, inside `elrond-go/cmd/storer2elastic` create `db/chainID` directory and merge dabatases from nodes in each shard (one node per shard). For a 2 shards + meta configuration, the directory should look like:
```
--db 
  --- chainID
     ----- Epoch_0
            ---- Shard_0
            ---- Shard_1
            ---- Shard_metachain
      ---- Epoch_1
            ---- Shard_0
            ---- Shard_1
            ---- Shard_metachain
     ....
```
3. create a new instance of elastic search (or index paths should be changed. example, from `blocks` -> `blocks2` -> this requires code changes)
4. adapt `config.toml` from storer2elastic if needed (update elastic search credentials and URL)
5. inside `storer2elastic/config` place the `nodesSetup.json` used when running the testnet
6. build
7. run (no flags are necessary, hopefully) 
8. this tool should index all the data and exit when finishing or encountering an error
9. for basic comparison, indexes can be manually checked. for advanced/automated comparison, the tool on `elrond-go-tools` can be used which just needs to be set up with the elastic search instances URLs and indexes paths
